internal/testutils/testserver.go

```diff
@@ -51,55 +51,87 @@ func (t TestServer) FaultyClient() *http.Client {
+func NewHTTPTestServer(t *testing.T, responses []ResponseDef) *TestServer {
 	testServer := &TestServer{}
 	handler := func(rw http.ResponseWriter, req *http.Request) {
 		testServer.calls++
-		// Special case: if only one response is specified, this will be reused
-		// for each and every call
-		if len(responses) == 1 {
-			if res, found := responses[0][req.Method]; found {
-				if res.ValidateRequestFunc != nil {
-					res.ValidateRequestFunc(req)
-				}
-				rw.WriteHeader(res.ResponseCode)
-				_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-			} else {
-				t.Errorf("unexpected HTTP method call: %s", req.Method)
-			}
-			return
+		if len(responses) <= testServer.calls-1 {
+			t.Errorf("Exceeded number of calls to test server (expected: %d)", len(responses))
 		}
 
-		if len(responses) <= testServer.calls-1 {
-			t.Fail()
-			t.Fatalf("Exceeded number of calls to test server (expected: %d)", len(responses))
+		responseDef := responses[testServer.calls-1]
+		methodFuncs := map[string]func(*testing.T, *http.Request) Response{
+			http.MethodGet:    responseDef.GET,
+			http.MethodPost:   responseDef.POST,
+			http.MethodPut:    responseDef.PUT,
+			http.MethodDelete: responseDef.DELETE,
 		}
 
-		if res, found := responses[testServer.calls-1][req.Method]; found {
-			if res.ValidateRequestFunc != nil {
-				res.ValidateRequestFunc(req)
-			}
-			rw.WriteHeader(res.ResponseCode)
-			_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
-		} else {
-			t.Errorf("unexpected HTTP method call: %s", req.Method)
+		handlerFunc, found := methodFuncs[req.Method]
+		if !found {
+			rw.WriteHeader(http.StatusMethodNotAllowed)
+			_, _ = rw.Write([]byte("Method not supported")) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+			return
 		}
+		if handlerFunc == nil {
+			t.Errorf("No %s method defined for server call nr. %d", req.Method, testServer.calls)
+			return
+		}
+
+		response := handlerFunc(t, req)
+		rw.WriteHeader(response.ResponseCode)
+		_, _ = rw.Write([]byte(response.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+		responseDef.Validate(t, req)
 	}
 	testServer.Server = httptest.NewServer(http.HandlerFunc(handler))
 	return testServer
 }
 
+type Response struct {
+	ResponseCode int
+	ResponseBody string
+}
+
+type ResponseDef struct {
+	GET             func(*testing.T, *http.Request) Response
+	PUT             func(*testing.T, *http.Request) Response
+	POST            func(*testing.T, *http.Request) Response
+	DELETE          func(*testing.T, *http.Request) Response
+	ValidateRequest func(*testing.T, *http.Request)
+}
+
+func (r ResponseDef) Get(t *testing.T, req *http.Request) Response {
```

Comment:
are these used anywhere? if I'm not mistaken you only use the struct fields directly?

Reply:
only `Validate` is used which provides a default passthrough if you didn't specify a ValidateRequest func.
forgot to call the others as well.

Reply:
Do you want to call them or remove them, though? 
With the current map and !found behavior there can't be a case that needs the != nil check these methods provide.
Could replace the map with a switch and method call maybe, but I don't know if that improves anything

Reply:
I would call them, it gives the possibility to add additional logic if needed (maybe never will be but it does not hurt if they are there, and for Validate it is actually already used)